File: go/test/moogle_api_web_features_games_get_get_game_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/moogleapi-sdk"
	"github.com/voxgig-sdk/moogleapi-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestMoogleApiWebFeaturesGamesGetGetGameEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.MoogleApiWebFeaturesGamesGetGetGame(nil)
		if ent == nil {
			t.Fatal("expected non-nil MoogleApiWebFeaturesGamesGetGetGameEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := moogle_api_web_features_games_get_get_gameBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "moogle_api_web_features_games_get_get_game." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		moogleApiWebFeaturesGamesGetGetGameRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.moogle_api_web_features_games_get_get_game", setup.data)))
		var moogleApiWebFeaturesGamesGetGetGameRef01Data map[string]any
		if len(moogleApiWebFeaturesGamesGetGetGameRef01DataRaw) > 0 {
			moogleApiWebFeaturesGamesGetGetGameRef01Data = core.ToMapAny(moogleApiWebFeaturesGamesGetGetGameRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = moogleApiWebFeaturesGamesGetGetGameRef01Data

		// LOAD
		moogleApiWebFeaturesGamesGetGetGameRef01Ent := client.MoogleApiWebFeaturesGamesGetGetGame(nil)
		moogleApiWebFeaturesGamesGetGetGameRef01MatchDt0 := map[string]any{
			"id": moogleApiWebFeaturesGamesGetGetGameRef01Data["id"],
		}
		moogleApiWebFeaturesGamesGetGetGameRef01DataDt0Loaded, err := moogleApiWebFeaturesGamesGetGetGameRef01Ent.Load(moogleApiWebFeaturesGamesGetGetGameRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		moogleApiWebFeaturesGamesGetGetGameRef01DataDt0LoadResult := core.ToMapAny(moogleApiWebFeaturesGamesGetGetGameRef01DataDt0Loaded)
		if moogleApiWebFeaturesGamesGetGetGameRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if moogleApiWebFeaturesGamesGetGetGameRef01DataDt0LoadResult["id"] != moogleApiWebFeaturesGamesGetGetGameRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func moogle_api_web_features_games_get_get_gameBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "moogle_api_web_features_games_get_get_game", "MoogleApiWebFeaturesGamesGetGetGameTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read moogle_api_web_features_games_get_get_game test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse moogle_api_web_features_games_get_get_game test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"moogle_api_web_features_games_get_get_game01", "moogle_api_web_features_games_get_get_game02", "moogle_api_web_features_games_get_get_game03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID": idmap,
		"MOOGLEAPI_TEST_LIVE":      "FALSE",
		"MOOGLEAPI_TEST_EXPLAIN":   "FALSE",
		"MOOGLEAPI_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["MOOGLEAPI_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["MOOGLEAPI_APIKEY"],
			},
			extra,
		})
		client = sdk.NewMoogleapiSDK(core.ToMapAny(mergedOpts))
	}

	live := env["MOOGLEAPI_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["MOOGLEAPI_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/moogle_api_web_features_games_get_get_game/MoogleApiWebFeaturesGamesGetGetGameEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { MoogleapiSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('MoogleApiWebFeaturesGamesGetGetGameEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when MOOGLEAPI_TEST_LIVE=TRUE.
  afterEach(liveDelay('MOOGLEAPI_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = MoogleapiSDK.test()
    const ent = testsdk.MoogleApiWebFeaturesGamesGetGetGame()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.MOOGLEAPI_TEST_LIVE
    for (const op of ['load']) {
      if (maybeSkipControl(t, 'entityOp', 'moogle_api_web_features_games_get_get_game.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let moogle_api_web_features_games_get_get_game_ref01_data = Object.values(setup.data.existing.moogle_api_web_features_games_get_get_game)[0] as any

    // LOAD
    const moogle_api_web_features_games_get_get_game_ref01_ent = client.MoogleApiWebFeaturesGamesGetGetGame()
    const moogle_api_web_features_games_get_get_game_ref01_match_dt0: any = {}
    moogle_api_web_features_games_get_get_game_ref01_match_dt0.id = moogle_api_web_features_games_get_get_game_ref01_data.id
    const moogle_api_web_features_games_get_get_game_ref01_data_dt0 = await moogle_api_web_features_games_get_get_game_ref01_ent.load(moogle_api_web_features_games_get_get_game_ref01_match_dt0)
    assert(moogle_api_web_features_games_get_get_game_ref01_data_dt0.id === moogle_api_web_features_games_get_get_game_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/moogle_api_web_features_games_get_get_game/MoogleApiWebFeaturesGamesGetGetGameTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = MoogleapiSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['moogle_api_web_features_games_get_get_game01','moogle_api_web_features_games_get_get_game02','moogle_api_web_features_games_get_get_game03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID': idmap,
    'MOOGLEAPI_TEST_LIVE': 'FALSE',
    'MOOGLEAPI_TEST_EXPLAIN': 'FALSE',
    'MOOGLEAPI_APIKEY': 'NONE',
  })

  idmap = env['MOOGLEAPI_TEST_MOOGLE_API_WEB_FEATURES_GAMES_GET_GET_GAME_ENTID']

  const live = 'TRUE' === env.MOOGLEAPI_TEST_LIVE

  if (live) {
    client = new MoogleapiSDK(merge([
      {
        apikey: env.MOOGLEAPI_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.MOOGLEAPI_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
